### DISCUSSION TITLE
Thermal printer support suggestion

### DISCUSSION THREAD

**Discussion by ralphrmartin** [arendst/Tasmota#16700] [Ideas]
It would be great if tasmota could add thermal printer support.
The way I see this working is that the printer would be attached to UART2, say, and upon receipt of an MQTT message on a specific topic, the contents of that message are printed on the thermal printer.
You might need to let the user resize the MQTT buffer to allow for long messages (e.g. a shopping receipt might be several thousand chars).
While there are various thermal printers, one like the Adafruit Thermal Printer might be a good starting point.

**Comment by barbudor** [comment]
For that it needs someone that has the device in hand
As I understand that your are talking of ESP32 and I see the is a CircuitPython driver for the printer, you may have a good chance of doing the port in Berry

**Comment by ralphrmartin** [comment]
Yes, I guess this would be for the ESP32. I suppose I could try to use Berry, but I just thought this might be a useful peripheral others might wish to use as well.

**Reply by barbudor** [reply]
Writing a driver in Berry doesn't prevent you for contributing it to Tasmota
There's a folder for that : https://github.com/arendst/Tasmota/tree/development/tasmota/berry/examples

**Reply by sfromis** [reply]
Whoever is going to do it, using Berry is a pretty good way of creating many drivers, as long as they do not require very exact timing of short intervals.

**Comment by ralphrmartin** [comment]
As far as I can see, Berry offers no way to change the size of the MQTT buffer.

**Reply by sfromis** [reply]
Tasmota as such does not have a feature to change MQTT max buffer size dynamically. If there was a command for that, it would also be available in Berry.

**Reply by barbudor** [reply]
You can override `MQTT_MAX_PACKET_SIZE` at compile time
Default value is 1200 which is more than enough to print a few lines
You don't have to send a whole DOC to the printer in a single MQTT message and it can be splitted as needed by the sender.

**Comment by ralphrmartin** [comment]
I have no control over the way the sender sends its messages. I need to provide a buffer size of 32000 bytes.

**Reply by Jason2866** [reply]
Why this environmentally unfriendly type of printer? Why when doing this not supporting a real printer? For [example](https://www.ebay.de/itm/OKI-Microline-ML3390-ML-3390-24-Nadel-Drucker-aus-Arztpraxis/304643444174?hash=item46ee2a0dce:g:6Q4AAOSwZqFjNInH&pageci=f50b63b6-7dec-46c8-bec6-37b97f991950&redirect=mobile). Printing is very easy with the old printer interface (parallel or rs232) Connected such a printer in the old days to a Sinclair ZX81.

**Reply by s-hadinger** [reply]
32k for a message is the high limit before you have serious memory allocation problems (unless you have PSRAM).

**Reply by barbudor** [reply]
@ralphrmartin
I don't know the details of your project but if you have no way on how the sender sends its message, how can you be sure that it will be able to follow Tasmota MQTT standard (`cmnd/%topic%/<COMMAND>`) ?

**Comment by ralphrmartin** [comment]
Jason2866: It is much more environmentally friendly not to waste a sheet of A4 paper for a small ticket, and not to use the much higher power requirements of an ink-jet or laser printer.
s-hadinger: esp32 has plenty of memory for a large buffer.
barbudor: Indeed, the external device will not follow the tasmota MQTT standard. Does this mean that tasmota devices cannot listen to arbitrary MQTT messages?

**Reply by barbudor** [reply]
If you do it for yourself in Berry, you can listen to whatever topic you want

If I was to implement a native driver for a thermal printer in Tasmota, I would follow the Tasmota standard like adding a commnd `ThermalPrint` that would answer to topic `cmnd/%topic%/ThermalPrint`

**Reply by Jason2866** [reply]
> Jason2866: It is much more environmentally friendly not to waste a sheet of A4 paper for a small ticket, and not to use the much higher power requirements of an ink-jet or laser printer. s-hadinger: esp32 has plenty of memory for a large buffer. barbudor: Indeed, the external device will not follow the tasmota MQTT standard. Does this mean that tasmota devices cannot listen to arbitrary MQTT messages?

It is more safe and friendly to use normal paper. There is still a high risk to get Thermopaper with Bisphenol A and Bisphenol S https://www.umweltbundesamt.de/themen/kassenbons-gehoeren-nicht-ins-altpapier#:~:text=Das%20Umweltbundesamt%20empfiehlt%20aus%20Vorsorgegr%C3%BCnden,von%20phenolfreien%20Thermopapieren%20unterscheiden%20kann.

**Reply by letshin** [reply]
Was any work done on this?

**Comment by ralphrmartin** [comment]
Not by me, and given the various comments of others, I doubt very much by anyone else.
